# OASIS Internship -Data Cleaning Project
## Titanic Dataset
### Objective
The objective is to clean and prepare the Titanic dataset for further analysis by identifying and handling missing value, duplicate records, inconsistent formatting, outliers, and incorrect data types.
### Cleaning Tasks
1. Produce a data quality report.
2. Handle missing data using appropriate strategies.
3. Identify and remove duplicate rows.
4. Detect and handle outliers.
5. standardize inconsistent formatting.
6. Correct data types.
7. Produce a before-versus-after cleaning summary.
8. Save the cleaned dataset as a new csv file.

In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [40]:
df = pd.read_csv("Titanic-Dataset.csv")
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [41]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 891
Number of columns: 12


In [42]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [44]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## Before Cleaning Assessment
Before cleaning the dataset, the number of rows, missing values, duplicate rows, and data types were recorded. These measurements will be compared with the dataset after cleaning.

In [45]:
before_rows = len(df)
before_nulls = df.isnull().sum().sum()
before_duplicates = df.duplicated().sum()

print("Rows before cleaning:",before_rows)
print("Total missing values:",before_nulls)
print("Duplicate rows:",before_duplicates)

Rows before cleaning: 891
Total missing values: 866
Duplicate rows: 0


## Data Quality Report - Missing Values
The number and percentage of missing values are calculated for every column to identify where missing data treatment is required.

In [46]:
null_report = pd.DataFrame({"Null_Count": df.isnull().sum(),"Null_Percentage":(df.isnull().sum()/ len(df) * 100).round(2)
                           })
null_report

,Null_Count,Null_Percentage
PassengerId,0,0.00
Survived,0,0.00
Pclass,0,0.00
Name,0,0.00
Sex,0,0.00
Age,177,19.87
SibSp,0,0.00
Parch,0,0.00
Ticket,0,0.00
Fare,0,0.00


## Data Qualiity Report - Duplicate Rows
The dataset is checked for completely duplicated rows. Duplicate records can cause repeated observations and may distort analysis.

In [47]:
duplicate_count = df.duplicated().sum()
print("Duplicate rows:",duplicate_count)

Duplicate rows: 0


In [48]:
dtype_report = pd.DataFrame({"Column": df.columns, "Current_Dtype":
                             df.dtypes.astype(str).values})
dtype_report

,Column,Current_Dtype
0,PassengerId,int64
1,Survived,int64
2,Pclass,int64
3,Name,object
4,Sex,object
5,Age,float64
6,SibSp,int64
7,Parch,int64
8,Ticket,object
9,Fare,float64


In [49]:
print("Sex values:")
print(df["Sex"].value_counts(dropna=False))

print("\nEmbarked values:")
print(df["Embarked"].value_counts(dropna=False))

Sex values:
Sex
male      577
female    314
Name: count, dtype: int64

Embarked values:
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


## Data Quality Report - Value Range Checks
Numerical columns are checked for unsuall or impossible values before cleaning. Values such as negative ages or negative fare would indicate potetial data quality problems.

In [50]:
print("Age:", df["Age"].min(), "to", df["Age"].max())
print("Fare:", df["Fare"].min(), "to", df["Fare"].max())
print("Survived:", sorted(df["Survived"].dropna().unique()))
print("Pclass:", sorted(df["Pclass"].dropna().unique()))
print("SibSp:", df["SibSp"].min(), "to", df["SibSp"].max())
print("Parch:", df["Parch"].min(), "to", df["Parch"].max())


Age: 0.42 to 80.0
Fare: 0.0 to 512.3292
Survived: [np.int64(0), np.int64(1)]
Pclass: [np.int64(1), np.int64(2), np.int64(3)]
SibSp: 0 to 8
Parch: 0 to 6


## Missing Data Handling - Age
The Age column contains 177 missing values.
Median imputation is used because age is a numerical variable and the median is less affected by extreme values than the mean. This approach allows us to retain passenger records rather than deleting them because of missing age information.

In [51]:
age_median = df["Age"].median()
print("median age:", age_median)
df["Age"] = df["Age"].fillna(age_median)

median age: 28.0


In [52]:
df["Age"].isnull().sum()

np.int64(0)

## Missing Data Handling - Cabin
The Cabin column contains 687 missing values, which represents a large proportion of the dataset.
Deleting all rows with missing cabin information would result in substantial information loss. Therefore, missing cabin values are replaced with "unknown". This preserves the passenger records while clearly indicating that cabin information was unavailable.

In [53]:
df["Cabin"] = df["Cabin"].fillna("unknown")

In [54]:
df["Cabin"].isnull().sum()

np.int64(0)

## Missing Data Handling - Embarked
The Embarked column contains only 2 missing values and is categorical.
Mode imputation is used because the mode represents the most frequently occuring category. This allows the two affected records to be retained without unnecessarily deleting rows.

In [55]:
embarked_mode = df["Embarked"].mode()[0]
print("Most common Embarked value:", embarked_mode)
df["Embarked"] = df["Embarked"].fillna(embarked_mode)

Most common Embarked value: S


In [56]:
df["Embarked"].isnull().sum()

np.int64(0)

In [57]:
missing_after = df.isnull().sum()

print(missing_after)
print("\nTotal missing values:", df.isnull().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Total missing values: PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64


## Duplicate Removal
The dataset was checked for duplicate rows. No duplicate rows were identified, so no records were removed. The original 891 passenger records are therefore retained.

In [58]:
duplicates_removed = df.duplicated().sum()
print("Duplicate rows identified:", duplicates_removed)
df = df.drop_duplicates().copy()
print("Rows after duplicate removal:", len(df))

Duplicate rows identified: 0
Rows after duplicate removal: 891


## Standardisation
Text fields are standardised by removing unnecessary leading and trailing spaces and applying consistent formatting.
Sex values are standardised to "Male" and "Female", while Embarked values are standardised to uppercase codes such as S, C, and Q.

In [59]:
# Remove unnecessary spaces
df["Name"] = df["Name"].str.strip()
df["Ticket"] = df["Ticket"].str.strip()
df["Cabin"] = df["Cabin"].str.strip()
# Standardise Sex
df["Sex"] =df["Sex"].str.strip().str.lower()

df["Sex"] = df["Sex"].replace({"m": "male", "f": "female"})
df["Sex"] =df["Sex"].str.title()

# Standardise Embarked
df["Embarked"] = df["Embarked"].str.strip().str.upper()

In [60]:
print(df["Sex"].unique())
print(df["Embarked"].unique())

['Male' 'Female']
['S' 'C' 'Q']


## Outlier Detection Using IQR
The Interquartile Range (IQR) method is used to identify potential outliers in numerical columns. The numerical columns examined are Age, SibSp, Parch, and Fare.
An observation identified as an outlier is not automatically considered an error. The values must be examined before deciding whether to remove, cap, or retain them.

In [61]:
def iqr_outlier_report(data, column):
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 - 1.5 * IQR

    outliers = data[
    (data[column] < lower_bound) |
    (data[column] > upper_bound)
        ]
    return {
    "column": column,
    "Q1": Q1,
    "Q3": Q3,
    "IQR": IQR,
    "Lower_Bound":lower_bound,
    "Upper_Bound": upper_bound,
    "Outlier_Count": len(outliers)
    }
    numeric_columns = ["Age", "SibSp", "Parch", "Fare"]
    outlier_results = []
    for column in numeric_columns:
        outlier_results.append(iqr_outlier_report(df, column)
                              )
        outlier_report = pd.DataFrame(outlier_results)
        outlier_report

## Outlier Handling Decision
The IQR method identified potential outliers in several numerical columns. The identified values were not automatically removed because statiscal outliers are not necessarily data_entry errors. In the Titanic dataset, unussually high fares can represent legitimate ticket purchases, while higher values in SibSp and Parch can represent genuine family relationships.
Age values within the observed range can also represent legitimate passenger ages. Therefore, the identified outliers are retained rather than removed or capped because there is insufficient evidence that they are erroneous values. Retaining them prevents the loss of potentially meaningful observations.

## Data Type Correction
Data types are corrected according to the meaning of each variable. Passengerid, Name, Ticket, and Cabin are treated as string fields.Pclass, Sex, and Embarked are categorical variables. Age, Fare are numerical variables, while Survived, SibSp, and Parch are integer variables.

In [62]:
df["PassengerId"] = df["PassengerId"].astype("string")
df["Survived"] = df["Survived"].astype("int64")
df["Pclass"] = df["Pclass"].astype("category")
df["Name"] = df["Name"].astype("string")
df["Sex"] = df["Sex"].astype("category")
df["Age"] = df["Age"].astype("float64")
df["SibSp"] = df["SibSp"].astype("int64")
df["Parch"] = df["Parch"].astype("int64")
df["Ticket"] = df["Ticket"].astype("string")
df["Fare"] = df["Fare"].astype("float64")
df["Cabin"] = df["Cabin"].astype("string")
df["Embarked"] = df["Embarked"].astype("category")

In [63]:
df.dtypes

PassengerId    string[python]
Survived                int64
Pclass               category
Name           string[python]
Sex                  category
Age                   float64
SibSp                   int64
Parch                   int64
Ticket         string[python]
Fare                  float64
Cabin          string[python]
Embarked             category
dtype: object

In [64]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    string  
 1   Survived     891 non-null    int64   
 2   Pclass       891 non-null    category
 3   Name         891 non-null    string  
 4   Sex          891 non-null    category
 5   Age          891 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    string  
 9   Fare         891 non-null    float64 
 10  Cabin        891 non-null    string  
 11  Embarked     891 non-null    category
dtypes: category(3), float64(2), int64(3), string(4)
memory usage: 65.8 KB


## Dtype Accuracy
The corrected data types are compared with the expected data types to determine the accuracy of the data type correction.

In [65]:
expected_dtypes = {
    "PassengerId": "string",
    "Survived": "int64",
    "Pclass": "category",
    "Name": "string",
    "Sex": "category",
    "Age": "float64",
    "SibSp": "int64",
    "Parch": "int64",
    "Ticket": "string",
    "Fare": "float64",
    "Cabin": "string",
    "Embarked": "category" 
}
dtype_accuracy = sum(
    str(df[column].dtype) == expected_dtype for column, expected_dtype in expected_dtypes.items())

total_columns = len(expected_dtypes)
print("correct data types:", dtype_accuracy,"/", total_columns)
print("Dtype accuracy:",
      round(dtype_accuracy / total_columns * 100, 2),
      "%"
     )
    
    


correct data types: 12 / 12
Dtype accuracy: 100.0 %


In [66]:
for column, expected_dtype in expected_dtypes.items():
    actual_dtype = str(df[column].dtype)
    if actual_dtype != expected_dtype:
        print(f"{column}: Actual = {actual_dtype}, Expected = {expected_dtype}")

In [67]:
print("Final Data Quality Check")
print("========================")

print("Rows:",df.shape[0])
print("columns:",df.shape[1])
print("Total missing value:",
      df.isnull().sum().sum())
print("Duplicate rows:",
     df.duplicated().sum())
print("Dtype accuracy:", f"{dtype_accuracy}/{total_columns}")

Final Data Quality Check
Rows: 891
columns: 12
Total missing value: 0
Duplicate rows: 0
Dtype accuracy: 12/12


In [68]:
after_rows = len(df)
after_nulls =df.isnull().sum().sum()
after_duplicates = df.duplicated().sum()

final_summary = pd.DataFrame({
    "Metric": ["Row count", "Total null count", "Duplicate row count", "Dtype accuracy"],
    "Before Cleaning": [before_rows, before_nulls, before_duplicates, "Required correction"],
    "After Cleaning": [after_rows, after_nulls, after_duplicates, f"{dtype_accuracy}/ {total_columns}( 100%)"]})
final_summary

,Metric,Before Cleaning,After Cleaning
0,Row count,891,891
1,Total null count,866,0
2,Duplicate row count,0,0
3,Dtype accuracy,Required correction,12/ 12( 100%)


## Save Cleaned Dataset
The cleaned dataset is saved as a new CSV file so that the original Titanic dataset remains unchanged.

In [69]:
df.to_csv("Titanic_Cleaned.csv", index = False)
print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


In [70]:
import os
print("Titanic_Cleaned.csv exists:",
      os.path.exists("Titanic_Cleaned.csv"))

Titanic_Cleaned.csv exists: True


In [72]:
cleaned_check = pd.read_csv("Titanic_Cleaned.csv")

print("Saved dataset shape:",cleaned_check.shape)
print("Missing values:", cleaned_check.isnull().sum().sum())
print("Duplicate rows:", cleaned_check.duplicated().sum())

Saved dataset shape: (891, 12)
Missing values: 0
Duplicate rows: 0


## Conclusion
The Titanic dataset was successfully cleaned and prepared for further analysis.
The original dataset contained 891 rows and 12 columns,with 866 missing values across the Age, Cabin, and Embarked columns. Median imputation was used for missing Age values, while missing Cabin values were replaced with "unknown". The two missing Embarked values were handled using mode imputation.
The dataset was checked for duplicate rows, and no duplicate records were identified. Text fields were standardised by removing unnecessary spaces and applying consistent formatting.
Potential numerical outliers were identified using the IQR method. The identified values were retained because they could represent legitimate passenger characteristics or ticket information rather than confirmed data entry error.
Data types were corrected according to the meaning of each varibles. Identifier and text fields were converted to string, categorical variables wre converted to category, and numerical variables were assigned appropriate numerical data types.
After cleaning, the dataset contained 891 rows and 12 columns, with zero missing values and zero duplicate rows. All 12 columns matched the expected data types.
The cleaned dataset was saved as "Titanic_Cleaned.csv", while the original dataset was preserved uchanged.